### ISSUE TITLE
First Field in useFieldArray Missing Upon Submission

### ISSUE THREAD

**Issue by thoughtassassin** [react-hook-form/react-hook-form#843]
**Describe the bug**
getValues does not accuately reflect what is on the UI.

**To Reproduce**
1. Go to codesanbox below.
2. Click add twice
3. Delete the top field that was added
4. When you click submit the remaining field doesn't show in the submission

**Codesandbox link**
https://codesandbox.io/s/react-hook-form-usefieldarray-5cgwv

**Expected behavior**
I am also logging getValues in the remove and add method. It seems like the first item in the array is never returned from getValues. I would expect once add or remove getValues will mirror what is displayed on the UI.

**Screenshots**
<img width="566" alt="Screen Shot 2020-01-15 at 9 10 02 AM" src="https://user-images.githubusercontent.com/1539057/72445214-fc7de880-3776-11ea-8ffe-ec35841e1df2.png">

**Desktop (please complete the following information):**
 - OS: MacOS
 - Browser Chrome
 - Version 79.0.3945.117

**Comment by bluebill1049** [comment]
I think you submit a wrong codesandbox link: https://codesandbox.io/s/x77o7v4kxz

**Comment by thoughtassassin** [comment]
https://codesandbox.io/s/react-hook-form-usefieldarray-5cgwv

**Comment by thoughtassassin** [comment]
Sorry about that!

**Comment by bluebill1049** [comment]
good find :) by the way `getValues()` is not going to work as you calling `append()` until `render`

**Comment by thoughtassassin** [comment]
Hmmm. I guess I would need to know what values are present after append. I have some logic to filter values passed to a select based upon previous selections.

**Comment by bluebill1049** [comment]
then you will probably need `watch()`, or call it during `useEffect()`

**Comment by bluebill1049** [comment]
getValues() is reading what's on the form, it can't performance when the item is not mounted yet.

**Comment by bluebill1049** [comment]
FYI: isolating the issue, looks like is `FormContext` related.

**Comment by thoughtassassin** [comment]
Awesome. I hope you find it!

**Comment by bluebill1049** [comment]
OMG found it! LOL you won't believe it! i spend 2hrs on this...

https://codesandbox.io/s/react-hook-form-usefieldarray-4wtf5

**Comment by bluebill1049** [comment]
couple things:

<button **type="button"** onClick={() => removeWithValues(index)}> Delete </button>

onClick={() => appendWithValues({ **envName**: "test" })}

**Comment by thoughtassassin** [comment]
Wow. Okay. That's embarrassing for me. Thanks for debugging that. As you suggested I'll use `watch` to get the current values.

**Comment by bluebill1049** [comment]
> Wow. Okay. That's embarrassing for me. Thanks for debugging that. As you suggested I'll use `watch` to get the current values.

I think you can use `getValues()` but do it in the useEffect, so the input is mounted and we can collect value from it.

`watch` is not the best for performance, because it's pretty much convert your input into controlled and re-render on each key-stroke due to value change. so i would still prefer to use getValues() but getValues is not watching the input FYI.

**Comment by thoughtassassin** [comment]
So the code in my project isn't apples to apples with what I had in my codepen. I am not sure I necessarily had the same errors in my code base. I was definitely seeing some strange behavior around the way the array was being managed.

 The important point is that I upgraded from 4.4.4 to 4.5.3 with the exact same code in my project, and now everything is working. Thank you so much for your help and hard work @bluebill1049.

I will look into the suggestion to do getValues in useEffect to make sure the component is mounted before calling getValues().

**Comment by bluebill1049** [comment]
`useFieldArray` is still young and fresh (new born 👶 ) hehe, so i do expect issues around them, but me and my team always there try to solve them :) It's great to hear things working from everyone too. ✨

**Comment by thoughtassassin** [comment]
Too soon Bill. I actually just upgraded my current stable codebase that doesn't useFieldArray to 4.5.3. I actually have to apply the changes to the branch that I upgraded to useFieldArray. What can I say? I got excited. I'll let you know if it is fixed or not in my codebase.

**Comment by bluebill1049** [comment]
cool let me know @thoughtassassin 👍

**Comment by thoughtassassin** [comment]
I think I know what happened in my code. I was trying to refactor my code to use useFieldArray and I ended up with a key on the custom component and then inside my component I had a key on the fieldset inside the custom component. 

Taking off the second key fixed the remove weirdness. Wow. I remember creating my own forms when I started working with React. I know dealing with keys and mounted elements was a nightmare.
